### DISCUSSION TITLE
Feature request: Allow users to add their own unique openai api keys via the web ui

### DISCUSSION THREAD

**Discussion by tkg61** [open-webui/open-webui#4477] [General]
**Is your feature request related to a problem? Please describe.**

Currently there is the only the ability to set openai api keys on the backend and not a per user basis. 

**Describe the solution you'd like**
Add the ability for users to set their own unique openai API key so that we can enable users to track their own individual API usage vs having to worry about it on the backend. We will have over 1000 users and would like to not have to use the same key for all of them.

**Describe alternatives you've considered**
N/A

**Additional context**

Having the ability to have a user set their own API key in the account settings panel would the most preferred route. with a small popup when the log in for the first time asking for the api key would be a nice to have.

**Comment by ino-senpai** [comment]
This feature will enable budget management for users and groups using Litellm.

**Comment by marleyjaffe** [comment]
Allowing per user or per group API keys is also what I am looking for to allow me to track and segregate costs. Would love to see this functionality!

**Comment by rgaricano** [comment]
Is done, enabling Direct Connections on admin/settings/connections permit user define OpenAI connections 
Admin
![OpwbWebui-conn](https://github.com/user-attachments/assets/032b0ea6-aa96-4ab9-9244-5c11e03ed22b)
User
![OpwbWebui-conn_user](https://github.com/user-attachments/assets/9636432d-83fe-4c86-a024-cf27b7e958a9)

**Comment by mkagit** [comment]
I believe most people will want this feature allowing the **administrator** to configure the endpoint and API key for the user, so that they can maintain control over groups, models within the Workplace, and so on.

**Comment by arnoldmq** [comment]
This will also allow features only available for backend models, like action buttons, preconfigured knowledge and params. 

In the mean while this workarounds is gonna have to do
use a function pipe
https://docs.openwebui.com/features/extensibility/plugin/functions/pipe#example-openai-proxy-pipe

Hard core the models list and when making the call get the api key from a users direct connection, this requires the user to create one direct connection with the api key.
